tests/phpunit/CRM/MembershipExtras/Hook/Pre/MembershipPaymentPlanProcessorTest.php

```diff
@@ -0,0 +1,147 @@
+    $_REQUEST['installments_frequency'] = 1;
+    $_REQUEST['installments_frequency_unit'] = 'month';
+
+    $contact = ContactFabricator::fabricate();
+    $params = [
+      'is_pay_later' => TRUE,
+      'skipLineItem' => 1,
+      'skipCleanMoney' => TRUE,
+      'receive_date' => '2020-06-27',
+      'contact_id' => $contact['id'],
+      'fee_amount' => 0,
+      'net_amount' => "1200",
+      'total_amount' => "1200",
+      'payment_instrument_id' => $this->getOptionValue('EFT', 'payment_instrument'),
+      'financial_type_id' => $this->getFinancialTypeID('Member Dues'),
+      'contribution_status_id' => 'Pending',
+      'currency' => NULL,
+      'is_test' => FALSE,
+      'campaign_id' => NULL,
+    ];
+    $paymentPlanCreator = new CRM_MembershipExtras_Hook_Pre_MembershipPaymentPlanProcessor($params);
+    $paymentPlanCreator->createPaymentPlan();
+    $recurringContribution = $paymentPlanCreator->getRecurringContribution();
+    $recurringContribution = civicrm_api3('ContributionRecur', 'get', [
+      'sequential' => 1,
+      'id' => $recurringContribution['id'],
+      'options' => ['limit' => 0],
+    ])['values'][0];
+
+    $this->assertEquals('27', $recurringContribution['cycle_day']);
+  }
+
+  public function testYearlyCycleDayIsCalculatedFromReceiveDate() {
+    $_REQUEST['installments'] = 1;
+    $_REQUEST['installments_frequency'] = 1;
+    $_REQUEST['installments_frequency_unit'] = 'year';
+
+    $contact = ContactFabricator::fabricate();
+    $params = [
+      'is_pay_later' => TRUE,
+      'skipLineItem' => 1,
+      'skipCleanMoney' => TRUE,
+      'receive_date' => '2020-02-01',
+      'contact_id' => $contact['id'],
+      'fee_amount' => 0,
+      'net_amount' => "1200",
+      'total_amount' => "1200",
+      'payment_instrument_id' => $this->getOptionValue('EFT', 'payment_instrument'),
+      'financial_type_id' => $this->getFinancialTypeID('Member Dues'),
+      'contribution_status_id' => 'Pending',
+      'currency' => NULL,
+      'is_test' => FALSE,
+      'campaign_id' => NULL,
+    ];
+    $paymentPlanCreator = new CRM_MembershipExtras_Hook_Pre_MembershipPaymentPlanProcessor($params);
+    $paymentPlanCreator->createPaymentPlan();
+    $recurringContribution = $paymentPlanCreator->getRecurringContribution();
+    $recurringContribution = civicrm_api3('ContributionRecur', 'get', [
+      'sequential' => 1,
+      'id' => $recurringContribution['id'],
+      'options' => ['limit' => 0],
+    ])['values'][0];
+
+    $this->assertEquals('32', $recurringContribution['cycle_day']);
+  }
+
+  public function testReceiveDateCalculationHookChangesReceiveDate() {
+    $_REQUEST['installments'] = 12;
+    $_REQUEST['installments_frequency'] = 1;
+    $_REQUEST['installments_frequency_unit'] = 'month';
+
+    $contact = ContactFabricator::fabricate();
+    $params = [
+      'is_pay_later' => TRUE,
+      'skipLineItem' => 1,
+      'skipCleanMoney' => TRUE,
+      'receive_date' => '2020-01-01',
```

Comment:
@MiyaNoctem  would be good if we can have dates in the test dynamic instead of specific date as the test would be awkward if we re-visit it in 2022?.

Reply:
@madeleinevironda  this would be for dates in line 21, 100 and 111